File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.py
class Solution:

    def lengthOfLongestSubstring(self, s: str) -> int:
        length = len(s)
        maxLength = 0
        charset = set()
        left = 0

        for right in range(length):
            if s[right] not in charset:
                charset.add(s[right])
                maxLength = max(maxLength, right-left+1)
            else:
                while s[right] in charset:
                    charset.remove(s[left])
                    left += 1
                charset.add(s[right])
        return maxLength

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int length = s.length();
        int maxLength = 0, left = 0;
        unordered_set<char> charset;

        for (int right = 0; right < length; right++) {
            if (charset.count(s[right]) == 0) {
                charset.insert(s[right]);
                maxLength = max(maxLength, right - left + 1);
            } else {
                while (charset.count(s[right])) {
                    charset.erase(s[left]);
                    left++;
                }
                charset.insert(s[right]);
            }
        }
        return maxLength;
    }
};